src/Compilers/VisualBasic/Portable/Symbols/NamedTypeSymbol.vb

```diff
@@ -1089,6 +1099,85 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
 
+        Friend Overridable ReadOnly Property AllRequiredMembers As ImmutableSegmentedDictionary(Of String, Symbol)
+            Get
+                TryCalculateRequiredMembersWithCache(Nothing)
+                Debug.Assert(Not _lazyRequiredMembers.IsDefault)
+                Return _lazyRequiredMembers
+            End Get
+        End Property
+
+        Friend Overridable ReadOnly Property HasRequiredMembersError As Boolean
+            Get
+                TryCalculateRequiredMembersWithCache(Nothing)
+                Debug.Assert(Not _lazyRequiredMembers.IsDefault)
+                Return _lazyRequiredMembers = s_requiredMembersErrorSentinel
+            End Get
+        End Property
+
+        Friend MustOverride ReadOnly Property HasAnyDeclaredRequiredMembers As Boolean
+
+        Private Function TryCalculateRequiredMembersWithCache(ByRef requiredMembersBuilder As ImmutableSegmentedDictionary(Of String, Symbol).Builder) As Boolean
+            Debug.Assert(requiredMembersBuilder Is Nothing)
+            Dim lazyRequiredMembers As ImmutableSegmentedDictionary(Of String, Symbol) = _lazyRequiredMembers
+
+            If Not lazyRequiredMembers.IsDefault Then
+                If lazyRequiredMembers = s_requiredMembersErrorSentinel Then
+                    Return False
+                Else
+                    requiredMembersBuilder = lazyRequiredMembers.ToBuilder()
+                    Return True
+                End If
+            End If
+
+            Dim success = TryCalculateRequiredMembers(requiredMembersBuilder)
+
+            Dim requiredMembers = If(success,
+                If(requiredMembersBuilder?.ToImmutable(), ImmutableSegmentedDictionary(Of String, Symbol).Empty),
+                s_requiredMembersErrorSentinel)
+
+            RoslynImmutableInterlocked.InterlockedInitialize(_lazyRequiredMembers, requiredMembers)
+
+            Return success
+        End Function
+
+        Private Function TryCalculateRequiredMembers(ByRef requiredMembersBuilder As ImmutableSegmentedDictionary(Of String, Symbol).Builder) As Boolean
+            If BaseTypeNoUseSiteDiagnostics?.TryCalculateRequiredMembersWithCache(requiredMembersBuilder) = False Then
+                Return False
+            End If
+
+            If (Not HasAnyDeclaredRequiredMembers) AndAlso requiredMembersBuilder Is Nothing Then
+                Return True
+            End If
+
+            If requiredMembersBuilder Is Nothing Then
+                requiredMembersBuilder = ImmutableSegmentedDictionary.CreateBuilder(Of String, Symbol)()
+            End If
+
+            For Each member In GetMembersUnordered()
+                Dim existingMember As Symbol = Nothing
+                ' Need to make sure that members from a base type weren't hidden by members from the current type. That is an error scenario
+                If requiredMembersBuilder.TryGetValue(member.Name, existingMember) Then
```

Comment:
>requiredMembersBuilder.TryGetValue(member.Name, existingMember [](http://example.com/codeflow?start=19&length=62)

It looks like we are looking in a collection that is getting mutated, and strictly speaking, doesn't represent the set of required members from the base. It is quite possible that this isn't going to make a difference for conventional scenarios. But it is hard to tell. For example, VB understands overloading of properties. I can easily imagine when parameter-less property is required, and parameterized property next to it is not. I think it will be better  (more predictable, more robust, etc.) to do exactly what we intend to do, i.e. check against required members from base, from its ```AllRequiredMembers```. And this comment is probably applicable to C# implementation as well.

Reply:
For example, is VB going to be able to properly handle instantiation  of ```C2``` from this example?
[https://sharplab.io/#v2:EYLgtghglgdgNAFxFANgHwAICYAEBhARgFgAoAb1JypwCcBTARwFcp6ATHDAZh1gRwAKBHGQDmdBAG4AzhMkBfSopKls+XCHzFylagG0MBAHQAlJjARQwdI3gD2YAA6o6NAMquAblADGdaUYAkjBsdAAergByENYAgggINFDATAh0ABQwMXR2AGbpQgCUhQC6ulTcvBY4CAAWUNJ6fDhhJTgAvAB8LZKkyqq4hgDspBQk1JwALDgAKv4I6YXlIssTAPodODB0AO7qiyJCmwTyvePUyvJAA==](https://sharplab.io/#v2:EYLgtghglgdgNAFxFANgHwAICYAEBhARgFgAoAb1JypwCcBTARwFcp6ATHDAZh1gRwAKBHGQDmdBAG4AzhMkBfSopKls+XCHzFylagG0MBAHQAlJjARQwdI3gD2YAA6o6NAMquAblADGdaUYAkjBsdAAergByENYAgggINFDATAh0ABQwMXR2AGbpQgCUhQC6ulTcvBY4CAAWUNJ6fDhhJTgAvAB8LZKkyqq4hgDspBQk1JwALDgAKv4I6YXlIssTAPodODB0AO7qiyJCmwTyvePUyvJAA==)
```
class C1
{
    required public int P1 {get;set;}
 }

class C2 : C1
{
    [System.Runtime.CompilerServices.IndexerNameAttribute(nameof(P1))]
    public int this[int x] => x;
}

class Program
{
    void Test()
    {
        _ = new C2() {P1 = 1};
    }
}
```

Now imagine that some language declared both properties in the same type.

Reply:
Here is an example how VB can consume both properties by name at the same time:
[https://sharplab.io/#v2:EYLgtghglgdgNAGxAN2HALiKCA+AFAV2ASgGMACAYQQgGdaqBGAWAChyPyAlAUwgBMA8jAQBPcngBOAewAOPSenF5G5AIIMAkjHQ8A5grYBRGPyo16bNoWJlzdBpQBMbTuW0ALBVHSOW7N05XQI5BZAUEaQEGXgFhMQkZeUVlRgAKAA91LR19BQBKbPdcg0lgkIBxHnRykM5edAJJGHIABlq3EzMqmoCOLsS5BSUOoL7yAGUicgAVHlp0NPzRkIARKDByLIBeCVUAaj20xmXxgangY1N7eiA](https://sharplab.io/#v2:EYLgtghglgdgNAGxAN2HALiKCA+AFAV2ASgGMACAYQQgGdaqBGAWAChyPyAlAUwgBMA8jAQBPcngBOAewAOPSenF5G5AIIMAkjHQ8A5grYBRGPyo16bNoWJlzdBpQBMbTuW0ALBVHSOW7N05XQI5BZAUEaQEGXgFhMQkZeUVlRgAKAA91LR19BQBKbPdcg0lgkIBxHnRykM5edAJJGHIABlq3EzMqmoCOLsS5BSUOoL7yAGUicgAVHlp0NPzRkIARKDByLIBeCVUAaj20xmXxgangY1N7eiA)
```
Public Class C1
    ReadOnly Property P1 As Integer
End Class

Public Class C2
    Inherits C1
        
        Overloads ReadOnly Property P1(x As Integer) As Integer
        Get
            Return 0
        End Get
    End Property
        
    Sub Test()
            Dim x = P1 + P1(1)
    End Sub
End Class
```